lib/chef/resource/aws_route53_hosted_zone.rb

```diff
@@ -0,0 +1,220 @@
+  attribute :name, kind_of: String, callbacks: { "domain name cannot end with a dot" => lambda { |n| n !~ /\.$/ } }
+
+  # The comment included in the CreateHostedZoneRequest element. String <= 256 characters.
+  attribute :comment, kind_of: String
+
+  attribute :aws_route53_zone_id, kind_of: String, aws_id_attribute: true
+
+  def record_sets(&block)
+    if block_given?
+      @record_sets_block = block
+    else
+      @record_sets_block
+    end
+  end
+
+  def aws_object
+    driver, id = get_driver_and_id
+    result = driver.route53_client.get_hosted_zone(id: id).hosted_zone if id rescue nil
+    if result
+      result.resource_record_sets = get_record_sets_from_aws(result.id).resource_record_sets
+      result
+    else
+      nil
+    end
+  end
+
+  # since this is used exactly once, it could plausibly be inlined in #aws_object.
+  def get_record_sets_from_aws(hosted_zone_id, opts={})
+    params = { hosted_zone_id: hosted_zone_id }.merge(opts)
+    driver.route53_client.list_resource_record_sets(params)
+  end
+end
+
+class Chef::Provider::AwsRoute53HostedZone < Chef::Provisioning::AWSDriver::AWSProvider
+
+  provides :aws_route53_hosted_zone
+  use_inline_resources
+
+  CREATE = "CREATE"
+  UPDATE = UPSERT = "UPSERT"
+  DELETE = "DELETE"
+  RRS_COMMENT = "Managed by chef-provisioning-aws"
+
+  attr_accessor :record_set_list
+
+  def make_hosted_zone_config(new_resource)
+    config = {}
+    # add :private_zone here once VPC validation is enabled.
+    [:comment].each do |attr|
+      value = new_resource.send(attr)
+      if value
+        config[attr] = value
+      end
+    end
+    config
+  end
+
+  def create_aws_object
+    converge_by "create new Route 53 zone #{new_resource}" do
+
+      # AWS stores some attributes off to the side here.
+      hosted_zone_config = make_hosted_zone_config(new_resource)
+
+      values = {
+        name: new_resource.name,
+        hosted_zone_config: hosted_zone_config,
+        caller_reference: "chef-provisioning-aws-#{SecureRandom.uuid.upcase}",  # required, unique each call
+      }
+
+      zone = new_resource.driver.route53_client.create_hosted_zone(values).hosted_zone
+      new_resource.aws_route53_zone_id(zone.id)
+
+      begin
+        record_set_resources = get_record_sets_from_resource(new_resource)
+
+        if record_set_resources
+          change_list = record_set_resources.map { |rs| rs.to_aws_change_struct(CREATE) }
+
+          new_resource.driver.route53_client.change_resource_record_sets(hosted_zone_id: new_resource.aws_route53_zone_id,
+                                                                         change_batch: {
+                                                                           comment: RRS_COMMENT,
+                                                                           changes: change_list,
+                                                                           })
+        end
+      rescue => ex
```

Comment:
Furthermore, this seems like a pretty good pattern when there are dependent, enclosed objects.  The convergent model says that we when the CCR is finished the aws object will match the resource or we will raise an error.  Because we couldn't match the resource, it seems really nice to try and clean up after ourselves instead of staying in a semi-broken state.  Although that isn't the worst because the user can just CCR again to converge.  Anyways, nice catch!